csrc/device_lower/analysis/thread_predicate.cpp

```diff
@@ -494,6 +504,192 @@ class RedundantUseAnalysis : BackwardVisitor {
+            break;
+          }
+        }
+      }
+    }
+  }
+  // Find all the root domains that are merged to the leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1*B2*B3]
+  std::vector<IterDomain*> getRootDomainsMergedToLeaf(IterDomain* ld) {
+    std::vector<IterDomain*> merged_root_domains;
+    std::vector<int> index_root_domain;
+    std::vector<IterDomain*> intermediate_domains = root_domain_;
+    auto all_exp = DependencyCheck::getAllExprsBetween(
+        {root_domain_.begin(), root_domain_.end()}, {ld});
+    for (auto expr : all_exp) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto outer_iter =
+            std::find(root_domain_.begin(), root_domain_.end(), merge->outer());
+        auto inner_iter =
+            std::find(root_domain_.begin(), root_domain_.end(), merge->inner());
+
+        if (outer_iter != root_domain_.end()) {
+          merged_root_domains.emplace_back(*outer_iter);
+          index_root_domain.emplace_back(
+              std::distance(root_domain_.begin(), outer_iter));
+        }
+        if (inner_iter != root_domain_.end()) {
+          merged_root_domains.emplace_back(*inner_iter);
+          index_root_domain.emplace_back(
+              std::distance(root_domain_.begin(), inner_iter));
+        }
+      } else {
+        // current analysis of predication is only valid if all the exprs
+        // between this lead domain and root domains are merge
+        return std::vector<IterDomain*>();
+      }
+    }
+    // The following sort is added because in NVFuserTest.FusionIssue2076_CUDA
+    // the order is [I3, I1, B2] while the correct order should be [I1, B2, I3]
+    size_t n_elements = merged_root_domains.size();
+    TORCH_INTERNAL_ASSERT(
+        n_elements, "The number of merged root domains should > 0");
+    std::vector<int> indices(n_elements);
+    std::iota(indices.begin(), indices.end(), 0);
+    std::sort(indices.begin(), indices.end(), [&](int a, int b) {
+      return index_root_domain[a] < index_root_domain[b];
+    });
+    std::vector<IterDomain*> merged_root_domains_sorted(n_elements);
+    for (size_t i = 0; i < n_elements; ++i) {
+      merged_root_domains_sorted[i] = merged_root_domains[indices[i]];
+    }
+
+    return merged_root_domains_sorted;
+  }
+
+  // Get the index of the leaf domain if we skip the broadcasted root domains
+  Val* getIndexWithoutBroadcast(
+      const std::vector<IterDomain*>& merged_root_domains,
+      ParallelType pt) {
+    const int ndim = (int)merged_root_domains.size();
+    // get the stride if we index the leaf domain using its root domains
+    std::vector<Val*> root_stride(ndim);
+    root_stride[ndim - 1] = GpuLower::current()->kernel()->oneVal();
+    for (int i = ndim - 2; i >= 0; i--) {
+      auto pre_crd = merged_root_domains[i + 1];
+      Val* pre_extent = pre_crd->isBroadcast()
+          ? concretized_broadcast_root_domains_.at(pre_crd)->extent()
```

Comment:
Because of [this](https://github.com/NVIDIA/Fuser/pull/11#discussion_r1199194319), this may not get the right extent.

Reply:
protected. if `that` happens, will not do this optimization.